river/expert/bandit.py

```diff
@@ -0,0 +1,269 @@
+            if not metric.works_with(model):
+                raise ValueError(f"{metric.__class__.__name__} metric can't be used to evaluate a " +
+                                 f'{model.__class__.__name__}')
+        super().__init__(models)
+        self.reward_scaler = copy.deepcopy(reward_scaler)
+        self.metric = copy.deepcopy(metric)
+        self.print_every = print_every
+
+        self.save_metric_values = save_metric_values
+        if save_metric_values:
+            self.metric_values: typing.List = []
+
+        self.save_percentage_pulled = save_percentage_pulled
+        if save_percentage_pulled:
+            self.store_percentage_pulled: typing.List = []
+
+        # Initializing bandits internals
+        self._n_arms = len(models)
+        self._n_iter = 0 # number of times learn_one is called
+        self._N = np.zeros(self._n_arms, dtype=np.int)
+        self._average_reward = np.zeros(self._n_arms, dtype=np.float)
+
+    @abc.abstractmethod
+    def _pull_arm(self):
+        pass
+
+    @abc.abstractmethod
+    def _update_arm(self, arm, reward):
+        pass
+
+    @abc.abstractmethod
+    def _pred_func(self, model):
+        pass
+
+    @property
+    def _best_model_idx(self):
+        # average reward instead of cumulated (otherwise favors arms which are pulled often)
+        return np.argmax(self._average_reward)
+
+    @property
+    def best_model(self):
+        return self[self._best_model_idx]
+
+    @property
+    def percentage_pulled(self):
+        percentages = self._N / sum(self._N)
+        return percentages
+
+    def predict_one(self, x):
+        best_arm = self._pull_arm()
+        y_pred = self._pred_func(self[best_arm])(x)
+        return y_pred
+
+    def learn_one(self, x, y):
+        chosen_arm = self._pull_arm()
+        chosen_model = self[chosen_arm]
+
+        y_pred = chosen_model.predict_one(x)
+        self.metric.update(y_pred=y_pred, y_true=y)
+        chosen_model.learn_one(x=x, y=y)
+
+        # Update bandit internals (common to all bandit)
+        reward = self._compute_scaled_reward(y_pred=y_pred, y_true=y)
+        self._n_iter += 1
+        self._N[chosen_arm] += 1
+        self._average_reward[chosen_arm] += (1.0 / self._N[chosen_arm]) * \
+                                            (reward - self._average_reward[chosen_arm])
+
+        # Specific update of the arm for certain bandit class
+        self._update_arm(chosen_arm, reward)
+
+        if self.print_every:
+            if (self._n_iter % self.print_every) == 0:
+                self._print_info()
+
+        if self.save_percentage_pulled:
+            self.store_percentage_pulled += [self.percentage_pulled]
+
+        if self.save_metric_values:
+            self.metric_values += [self.metric._eval(y_pred, y)]
+
+        return self
+
+    def add_models(self, new_models):
+        if not isinstance(new_models, list):
+            raise TypeError("Argument `new_models` must be of a list")
```

Comment:
It's not Pythonic to check like this, just add a type hint to `new_models` and expect the user to read the documentation :)

Reply:
IMO the most pythonic way is to use try-catch cf EAFP (easier to ask for forgiveness than permission, see https://docs.python.org/3.6/glossary.html).

But if you prefer I can just add the type hint !

Reply:
I end up adding type hint